### PR TITLE
Inline stylesheets on load

### LINKED ISSUE #764: Dynamically loaded CSS files are not inlined in SPA(动态加载的css文件没有进行内联)
## 问题：动态加载的css文件没有进行内联

s1

下面是在播放的时候去请求了css文件：
s2

## 我的使用场景
我的场景是需要记录网站每一个版本的图片，我的处理方式是采用图片内容hash处理，然后hash值保存到图片url中，然后将这些图片单独保存到图片库。
在播放的时候，我会将event中的图片地址替换为图片库对应的地址，还原出对应版本的图片。

由于动态css文件没有进行内联，所以css文件中的图片地址没有保存到event中，我就无法在播放的时候动态修改这些图片的地址。

### PR DESCRIPTION
**Pull request by Juice10**
Fixes #764

### CODE REVIEW

File: packages/rrweb/test/record.test.ts [hunk 1]
```diff
@@ -34,9 +34,12 @@ const setup = function (this: ISuite, content: string): ISuite {
   const ctx = {} as ISuite;
 
   beforeAll(async () => {
-    ctx.browser = await launchPuppeteer();
+    ctx.browser = await launchPuppeteer({
+      devtools: true,
+    });
 
     const bundlePath = path.resolve(__dirname, '../dist/rrweb.min.js');
+    // const bundlePath = path.resolve(__dirname, '../dist/rrweb-all.js');
```
Comment (Juice10) [inline packages/rrweb/test/record.test.ts:42]:
```suggestion
```

File: packages/rrweb/test/record.test.ts [hunk 2]
```diff
@@ -57,6 +60,7 @@ const setup = function (this: ISuite, content: string): ISuite {
   });
 
   afterEach(async () => {
+    // await ctx.page.waitForTimeout(60000);
```
Comment (Juice10) [inline packages/rrweb/test/record.test.ts:63]:
```suggestion
```

File: packages/rrweb/test/record.test.ts [hunk 3]
```diff
@@ -68,7 +72,8 @@ const setup = function (this: ISuite, content: string): ISuite {
 };
 
 describe('record', function (this: ISuite) {
-  jest.setTimeout(10_000);
+  jest.setTimeout(180_000);
+  // jest.setTimeout(10_000);
```
Comment (Juice10) [inline packages/rrweb/test/record.test.ts:76]:
```suggestion
  jest.setTimeout(10_000);
```

File: packages/rrweb/scripts/repl.js [hunk 4]
```diff
@@ -119,7 +119,7 @@ void (async () => {
     await page.evaluate(`;${code}
       window.__IS_RECORDING__ = true
       rrweb.record({
-        emit: event => window._replLog(event),
+        emit: event => console.log(event),
```
Comment (Juice10) [inline packages/rrweb/scripts/repl.js:122]:
```suggestion
        emit: event => window._replLog(event),
```

File: packages/rrweb/test/record.test.ts [hunk 5]
```diff
@@ -381,10 +383,90 @@ describe('record', function (this: ISuite) {
...
+    await ctx.page.waitForResponse(corsStylesheetURL); // wait for stylesheet to be loaded
+    await waitForRAF(ctx.page); // wait for rrweb to emit events
+
+    assertSnapshot(ctx.events);
+  });
 });
 
 describe('record iframes', function (this: ISuite) {
-  jest.setTimeout(10_000);
+  jest.setTimeout(180_000);
```
Comment (Juice10) [inline packages/rrweb/test/record.test.ts:469]:
```suggestion
  jest.setTimeout(10_000);
```

File: packages/rrweb-snapshot/src/snapshot.ts [hunk 6]
```diff
@@ -377,6 +378,40 @@ function onceIframeLoaded(
...
 
+function isStylesheetLoaded(link: HTMLLinkElement) {
+  if (!link.getAttribute('href')) return true; // nothing to load
+  return link.sheet !== null;
+}
+
+function onceStylesheetLoaded(
+  link: HTMLLinkElement,
+  listener: () => unknown,
+  iframeLoadTimeout: number,
```
Comment (Yuyz0112) [inline packages/rrweb-snapshot/src/snapshot.ts:389]: shall we name this to styleSheetLoadTimeout?

**Comment by Yuyz0112** [comment]
there are some conflicts now

File: packages/rrweb-snapshot/src/snapshot.ts (hunk 6 above)
Comment (Juice10) [reply packages/rrweb-snapshot/src/snapshot.ts:389]: 🤦

yes, we should

File: packages/rrweb-snapshot/src/snapshot.ts [hunk 7]
```diff
@@ -1059,18 +1103,71 @@ export function serializeNodeWithId(
...
+        if (onStylesheetLoad) {
+          const serializedLinkNode = serializeNodeWithId(n, {
+            doc,
+            mirror,
+            blockClass,
+            blockSelector,
+            maskTextClass,
+            maskTextSelector,
+            skipChild: false,
+            inlineStylesheet,
```
Comment (eoghanmurray) [inline packages/rrweb-snapshot/src/snapshot.ts:1141]: @Juice10 I think this whole `if` could have been false, if inlineStylesheet is false?

Comment (Juice10) [reply packages/rrweb-snapshot/src/snapshot.ts:1141]: Yes that's very much possible. To be honest it's a while ago so I'm not sure if there was a deliberate intention to call this even if inlineStylesheet was false, but looking back at it I don't think so

Comment (eoghanmurray) [reply packages/rrweb-snapshot/src/snapshot.ts:1141]: It seems like the only upshot of `serializeNodeWithId` here is to extract out the `_cssText` from the .sheet (as this node has already been serialized).  in the rrweb callback that's all that is used.  Would it be fair to refactor based on that insight?  I'll be doing some work with assets related to this.

**Pull request merged**